registration/include/pcl/registration/incremental_icp.h

```diff
@@ -0,0 +1,130 @@
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef PCL_REGISTRATION_INCREMENTAL_ICP_H_
+#define PCL_REGISTRATION_INCREMENTAL_ICP_H_
+
+#include <pcl/point_cloud.h>
+#include <pcl/registration/registration.h>
+
+namespace pcl {
+  namespace registration {
+
+    /** \brief Incremental @ref IterativeClosestPoint class
+      *
+      * This class provides a way to register a stream of clouds where each cloud will be aligned to the previous cloud.
+      *
+      * \code
+      * IterativeClosestPoint<PointXYZ,PointXYZ>::Ptr icp (new IterativeClosestPoint<PointXYZ,PointXYZ>);
+      * icp->setMaxCorrespondenceDistance (0.05);
+      * icp->setMaximumIterations (50);
+      *
+      * IncrementalICP<PointXYZ> iicp;
+      * iicp.setICP (icp);
+      *
+      * while (true){
+      *   PointCloud<PointXYZ>::Ptr cloud (new PointCloud<PointXYZ>);
+      *   read_cloud (*cloud);
+      *   iicp.registerCloud (cloud);
+      *
+      *   PointCloud<PointXYZ>::Ptr tmp (new PointCloud<PointCloudXY>);
+      *   transformPointCloud (*cloud, *tmp, iicp.getAbsoluteTransform ());
+      *   write_cloud (*tmp);
+      * }
+      * \endcode
+      *
+      * \author Michael 'v4hn' Goerner
+      * \ingroup registration
+      */
+    template <typename PointT, typename Scalar = float>
+    class IncrementalICP {
+      public:
+        typedef typename pcl::PointCloud<PointT>::Ptr PointCloudPtr;
+        typedef typename pcl::PointCloud<PointT>::ConstPtr PointCloudConstPtr;
+
+        typedef typename pcl::Registration<PointT,PointT,Scalar>::Ptr RegistrationPtr;
+        typedef typename pcl::Registration<PointT,PointT,Scalar>::Matrix4 Matrix4;
+
+        IncrementalICP ();
+
+        /** \brief Empty destructor */
+        virtual ~IncrementalICP () {}
+
+        /** \brief Register new point cloud incrementally
+          * \note You have to set a valid registration object with @ref setICP before using this
+          * \note The class doesn't copy cloud. If you afterwards change cloud, that will affect this class.
+          * \param[in] cloud point cloud to register
+          * \param[in] delta_estimate estimated transform between last registered cloud and this one
+          * \return true if ICP converged
+          */
+        bool
+        registerCloud (const PointCloudConstPtr& cloud, const Matrix4& delta_estimate = Matrix4::Identity ());
+
+        /** \brief Get estimated transform between the last two registered clouds */
+        inline Matrix4
```

Comment:
AFAIK compilers may or may not respect the `inline` keyword, it is only a hint to the compiler.

I didn't look at how the interface is defined in other registration classes. Since there is such a big body of code that is implemented this way, and since there are no problems reported so far, I'd say let's keep this way.